README.md
# Lolcommits Yammer

[![Gem](https://img.shields.io/gem/v/lolcommits-yammer.svg?style=flat)](http://rubygems.org/gems/lolcommits-yammer)
[![Travis](https://img.shields.io/travis/com/lolcommits/lolcommits-yammer/master.svg?style=flat)](https://travis-ci.com/lolcommits/lolcommits-yammer)
[![Depfu](https://img.shields.io/depfu/lolcommits/lolcommits-yammer.svg?style=flat)](https://depfu.com/github/lolcommits/lolcommits-yammer)
[![Maintainability](https://api.codeclimate.com/v1/badges/dc8b0801920bffbecf9f/maintainability)](https://codeclimate.com/github/lolcommits/lolcommits-yammer/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/dc8b0801920bffbecf9f/test_coverage)](https://codeclimate.com/github/lolcommits/lolcommits-yammer/test_coverage)

[lolcommits](https://lolcommits.github.io/) takes a snapshot with your
webcam every time you git commit code, and archives a lolcat style image
with it. Git blame has never been so much fun!

This plugin automatically posts a new message to your Yammer account
with the captured lolcommit, your commit message and a 'Lolcommits'
topic tag. Something like this:

![example
commit](https://github.com/lolcommits/lolcommits-yammer/raw/master/assets/images/example-commit.png)

## Requirements

* Ruby >= 2.4
* A webcam
* [ImageMagick](http://www.imagemagick.org)
* [ffmpeg](https://www.ffmpeg.org) (optional) for animated gif capturing
* A [Yammer](http://yammer.com) account

## Installation

After installing the lolcommits gem, install this plugin with:

    $ gem install lolcommits-yammer

Then configure to enable with:

    $ lolcommits --config -p yammer
    # set enabled to `true`
    # confirm access for this plugin at yammer.com (link opens automatically)

That's it! Your next lolcommit will be posted to Yammer. To disable
uninstall this gem or use:

    $ lolcommits --config -p yammer
    # and set enabled to `false`

To revoke plugin permissions at Yammer, visit 'Edit Settings' -> 'My
Applications' and select 'Revoke Access' for the 'Lolcommits Yammer'
app.

## Development

Check out this repo and run `bin/setup`, this will install all
dependencies and generate docs. Use `bundle exec rake` to run all tests
and generate a coverage report.

You can also run `bin/console` for an interactive prompt, allowing you
to experiment with the gem code.

## Tests

MiniTest is used for testing. Run the test suite with:

    $ rake test

## Docs

Generate docs for this gem with:

    $ rake rdoc

## Troubles?

If you think something is broken or missing, please raise a new
[issue](https://github.com/lolcommits/lolcommits-yammer/issues). Take a
moment to check it hasn't been raised in the past (and possibly closed).

## Contributing

Bug [reports](https://github.com/lolcommits/lolcommits-yammer/issues)
and [pull
requests](https://github.com/lolcommits/lolcommits-yammer/pulls) are
welcome on GitHub.

When submitting pull requests, remember to add tests covering any new
behaviour, and ensure all tests are passing on [Travis
CI](https://travis-ci.com/lolcommits/lolcommits-yammer). Read the
[contributing
guidelines](https://github.com/lolcommits/lolcommits-yammer/blob/master/CONTRIBUTING.md)
for more details.

This project is intended to be a safe, welcoming space for
collaboration, and contributors are expected to adhere to the
[Contributor Covenant](http://contributor-covenant.org) code of conduct.
See
[here](https://github.com/lolcommits/lolcommits-yammer/blob/master/CODE_OF_CONDUCT.md)
for more details.

## License

The gem is available as open source under the terms of
[LGPL-3](https://opensource.org/licenses/LGPL-3.0).

## Links

* [Travis CI](https://travis-ci.com/lolcommits/lolcommits-yammer)
* [Code Climate](https://codeclimate.com/github/lolcommits/lolcommits-yammer)
* [Test Coverage](https://codeclimate.com/github/lolcommits/lolcommits-yammer/coverage)
* [RDoc](http://rdoc.info/projects/lolcommits/lolcommits-yammer)
* [Issues](http://github.com/lolcommits/lolcommits-yammer/issues)
* [Report a bug](http://github.com/lolcommits/lolcommits-yammer/issues/new)
* [Gem](http://rubygems.org/gems/lolcommits-yammer)
* [GitHub](https://github.com/lolcommits/lolcommits-yammer)


CHANGELOG.md
# ChangeLog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][KeepAChangelog] and this
project adheres to [Semantic Versioning][Semver].

## [Unreleased]

- Your contribution here!

## [0.5.0] - 2020-01-24
### Removed
- Support for Ruby < 2.4 (older rubies no longer supported)

## [0.4.0] - 2019-08-25
### Changed
- Require an up-to-date `rest-client` (`>= 2.1.0`)

## [0.3.0] - 2019-05-20
### Changed
- lolcommits gem is a runtime dependency
- Use `lolcommit_path` (instead of `main_image`)
- Requires at least lolcommits >= `0.14.2`
- Updated README and gemspec

### Removed
- Support for lolcommits < `0.14.2`

## [0.2.0] - 2019-04-24
### Removed
- Support for Ruby < 2.3 (older rubies no longer supported)

### Added
- `frozen_string_literal: true` to all ruby files

## [0.1.1] - 2018-05-24
### Changed
- Updated gemspec meta data links.

## [0.1.0] - 2018-03-27
### Changed
- Remove runner_order method
- Require at least lolcommits 0.12.0

## [0.0.6] - 2018-02-05
### Removed
- Dropped support for Ruby 2.0

## [0.0.5] - 2018-01-22
### Changed
- Bump Travis Rubies
- Refactor plugin config
- Require at least lolcommits 0.10.0

## [0.0.4] - 2017-12-06
### Changed
- remove vendored fonts, gh-pages

## [0.0.3] - 2017-11-22
### Changed
- ditch yam gem and net/http, use rest-client

## [0.0.2] - 2017-11-10
### Changed
- Don't log webrick

## [0.0.1] - 2017-11-10
### Changed
- Initial release

[Unreleased]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.5.0...HEAD
[0.5.0]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.1.1...v0.2.0
[0.1.1]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.0.6...v0.1.0
[0.0.6]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/lolcommits/lolcommits-yammer/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/lolcommits/lolcommits-yammer/compare/30cac66...v0.0.1
[KeepAChangelog]: http://keepachangelog.com/en/1.0.0/
[Semver]: http://semver.org/spec/v2.0.0.html


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

# lolcommits gem
require 'lolcommits'

# lolcommit test helpers
require 'lolcommits/test_helpers/git_repo'
require 'lolcommits/test_helpers/fake_io'

if ENV['COVERAGE']
  require 'simplecov'
end

# plugin gem test libs
require 'lolcommits/yammer'
require 'minitest/autorun'

# swallow all debug output during test runs
def debug(msg); end

# do not launch URLs
class Lolcommits::CLI::Launcher
  def self.open_url(url);  end
end


test/lolcommits/plugin/yammer_test.rb
# frozen_string_literal: true

require "json"
require "test_helper"
require 'webmock/minitest'

describe Lolcommits::Plugin::Yammer do
  include Lolcommits::TestHelpers::GitRepo
  include Lolcommits::TestHelpers::FakeIO

  describe "with a runner" do
    def runner
      @runner ||= Lolcommits::Runner.new(
        lolcommit_path: Tempfile.new('lolcommit.jpg').path
      )
    end

    def plugin
      @plugin ||= Lolcommits::Plugin::Yammer.new(runner: runner)
    end

    def valid_enabled_config
      {
        enabled: true,
        access_token: "oV4MuwnNKql3ebJMAYZRaD"
      }
    end

    describe "#enabled?" do
      it "is false by default" do
        _(plugin.enabled?).must_equal false
      end

      it "is true when configured" do
        plugin.configuration = valid_enabled_config
        _(plugin.enabled?).must_equal true
      end
    end

    describe "run_capture_ready" do
      before do
        plugin.configuration = valid_enabled_config
        commit_repo_with_message("first commit!")
      end

      after { teardown_repo }

      it "posts lolcommit to Yammer with commit message" do
        in_repo do
          stub_request(:post, create_message_api_url).to_return(status: 201)
          output = fake_io_capture { plugin.run_capture_ready }

          _(output).must_equal "Posting to Yammer ... done!\n"
          assert_requested :post, create_message_api_url, times: 1, headers: {
            'Authorization' => 'Bearer oV4MuwnNKql3ebJMAYZRaD',
            'Content-Type'  => /multipart\/form-data/
          } do |req|
            _(req.body).must_match(/Content-Disposition: form-data;.+name="attachment1"; filename="lolcommit.jpg.+"/)
          end
        end
      end

      it "reports an error if posting to yammer fails" do
        in_repo do
          stub_request(:post, create_message_api_url).to_return(status: 503)
          output = fake_io_capture { plugin.run_capture_ready }

          _(output.split("\n")).must_equal(
            [
              "Posting to Yammer ... failed :(",
              "Yammer error: 503 Service Unavailable",
              "Try a lolcommits capture with `--debug` and check for errors: `lolcommits -c --debug`"
            ]
          )
        end
      end
    end

    describe "configuration" do

      describe "#valid_configuration?" do
        it "returns false when not configured correctly" do
          _(plugin.valid_configuration?).must_equal false
        end

        it "returns true when configured" do
          plugin.configuration = valid_enabled_config
          _(plugin.valid_configuration?).must_equal true
        end
      end

      describe "configuring with Yammer Oauth" do
        before do
          # allow requests to localhost for this test
          WebMock.disable_net_connect!(allow_localhost: true)
        end

        after do
          WebMock.disable_net_connect!
        end

        it "aborts if Yammer Oauth is denied" do
          configured_plugin_options = {}
          fake_authorize_step

          output = fake_io_capture(inputs: %w(true)) do
            configured_plugin_options = plugin.configure_options!
          end

          _(output.split("\n").last).must_equal(
            "Aborting.. Plugin disabled since Yammer Oauth was denied"
          )

          _(configured_plugin_options).must_equal({ enabled: false })
        end

        it "configures successfully with a Yammer Oauth access token" do
          configured_plugin_options = {}
          yammer_oauth_token = "yam-oauth-token"
          yammer_oauth_code  = "yam-oauth-code"
          klass              = plugin.class

          stub_request(:post, klass::ACCESS_TOKEN_URL).with(
            body: {
              "client_id"     => klass::OAUTH_CLIENT_ID,
              "client_secret" => klass::OAUTH_CLIENT_SECRET,
              "code"          => [yammer_oauth_code]
            }
          ).to_return(
            status: 200,
            body: {
              "access_token" => {
                "token" => yammer_oauth_token
              }
            }.to_json
          )

          fake_authorize_step("?code=#{yammer_oauth_code}")

          fake_io_capture(inputs: %w(true)) do
            configured_plugin_options = plugin.configure_options!
          end

          _(configured_plugin_options).must_equal({
            enabled: true,
            access_token: "yam-oauth-token"
          })
        end
      end
    end
  end

  private

  def create_message_api_url
    plugin.class::MESSAGES_API_URL
  end

  # fake click for the authorize step in Yammer, by hitting local webrick server
  # loops repeating request until the server responds 200 OK
  def fake_authorize_step(redirect_params = nil)
    fork do
      res = nil
      while !res || res.code != "200"
        uri = URI("#{plugin.class::OAUTH_REDIRECT_URL}/#{redirect_params}")
        res = Net::HTTP.get_response(uri) rescue nil
        sleep 0.1
      end
    end
  end
end


lib/lolcommits/plugin/yammer.rb
# frozen_string_literal: true

require 'lolcommits/plugin/base'
require 'lolcommits/cli/launcher'
require 'webrick'
require 'rest-client'
require 'cgi'

module Lolcommits
  module Plugin
    class Yammer < Base

      MESSAGES_API_URL    = "https://www.yammer.com/api/v1/messages".freeze
      ACCESS_TOKEN_URL    = 'https://www.yammer.com/oauth2/access_token.json'.freeze
      OAUTH_CLIENT_ID     = 'abbxXRgeSagk9GtiWW9rFw'.freeze
      OAUTH_CLIENT_SECRET = 'gHVw5Ekyy2mWOWsBzrZPs5EPnR6s04RibApcbuy10'.freeze
      OAUTH_REDIRECT_PORT = 5429
      OAUTH_REDIRECT_URL  = "http://localhost:#{OAUTH_REDIRECT_PORT}".freeze

      ##
      # Returns true if the plugin has been configured correctly.
      # The access_token must be set.
      #
      # @return [Boolean] true/false
      #
      def valid_configuration?
        !configuration[:access_token].nil?
      end

      ##
      # Prompts the user to configure the plugin.
      #
      # If the enabled option is set we attempt to fetch an Oauth token
      # via Yammers' Oauth 2 Server Side flow.
      #
      # https://developer.yammer.com/docs/oauth-2#server-side-flow
      #
      # @return [Hash] the configured plugin options
      #
      def configure_options!
        options = super
        if options[:enabled]
          oauth_access_token = fetch_access_token
          if oauth_access_token
            options.merge!(access_token: oauth_access_token)
          else
            puts "Aborting.. Plugin disabled since Yammer Oauth was denied"
            options[:enabled] = false
          end
        end
        options
      end

      ##
      # Post-capture hook, runs after lolcommits captures a snapshot.
      # Posts the lolcommit file to Yammer with a commit message
      # postfixed by a #lolcommits topic/hashtag.
      #
      # @return [Boolean] true/false indicating posting was successful
      #
      def run_capture_ready
        print "Posting to Yammer ... "
        response = RestClient.post(
          "https://www.yammer.com/api/v1/messages",
          { body: yammer_message, attachment1: File.new(runner.lolcommit_path) },
          { 'Authorization' => "Bearer #{configuration[:access_token]}" }
        )

        if response.code != 201
          debug response.body.inspect
          raise "Invalid response code (#{response.code})"
        end

        print "done!\n"
        true
      rescue StandardError => e
        print "failed :(\n"
        puts "Yammer error: #{e.message}"
        puts "Try a lolcommits capture with `--debug` and check for errors: `lolcommits -c --debug`"
        false
      end


      private

      def yammer_message
        "#{runner.message} #lolcommits"
      end

      def fetch_access_token
        puts "\nOpening this url to authorize lolcommits-yammer:"
        puts authorize_url
        open_url(authorize_url)
        puts "\nLaunching local webbrick server to complete the OAuth process ...\n"
        @oauth_code = nil
        begin
          trap('INT') { local_server.shutdown }
          trap('SIGTERM') { local_server.shutdown }
          local_server.mount_proc '/', server_callback
          local_server.start

          if @oauth_code
            debug "Requesting Yammer OAuth Token with code: #{@oauth_code}"
            oauth_response = RestClient.post(ACCESS_TOKEN_URL,
              {
                'client_id'     => OAUTH_CLIENT_ID,
                'client_secret' => OAUTH_CLIENT_SECRET,
                'code'          => @oauth_code
              }
            )

            if oauth_response.code.to_i == 200
              return JSON.parse(oauth_response.body)['access_token']['token']
            end
          end
          return nil
        rescue WEBrick::HTTPServerError
          raise "Do you have something running on port #{OAUTH_REDIRECT_PORT}? Please turn it off to complete the oauth process"
        end
      end

      def authorize_url
        "https://www.yammer.com/oauth2/authorize?client_id=#{OAUTH_CLIENT_ID}&response_type=code&redirect_uri=#{OAUTH_REDIRECT_URL}"
      end

      def open_url(url)
        Lolcommits::CLI::Launcher.open_url(url)
      end

      def local_server
        @local_server ||= WEBrick::HTTPServer.new(
          Port: OAUTH_REDIRECT_PORT,
          Logger: null_logger,
          AccessLog: null_logger
        )
      end

      def null_logger
        WEBrick::Log.new(nil, -1)
      end

      def oauth_response(heading)
        <<-RESPONSE
          <html>
            <head>
              <style>
              body {
                background-color: #36465D;
                text-align: center;
              }

              a { color: #529ecc; text-decoration: none; }
              a img { border: none; }

              img {
                width: 100px;
                margin-top: 100px;
              }

              div {
                margin: 20px auto;
                font: normal 16px "Helvetica Neue", "HelveticaNeue", Helvetica, Arial, sans-serif;
                padding: 20px 40px;
                background: #FEFEFE;
                width: 50%;
                border-radius: 10px;
                color: #757575;
              }

              h1 {
                font-size: 18px;
              }
              </style>
            </head>
            <body>
              <a href="https://lolcommits.github.io">
                <img src="https://lolcommits.github.io/assets/img/logo/lolcommits_logo_400px.png" alt="lolcommits" width="100px" />
              </a>
              <div>
                <h1>#{heading}</h1>
              </div>
            </body>
          </html>
        RESPONSE
      end

      def server_callback
        proc do |req, res|
          local_server.stop
          local_server.shutdown

          query = req.request_uri.query
          query = CGI.parse(req.request_uri.query) if query

          if query && !query['code'].empty?
            @oauth_code = query['code']
            res.body = oauth_response("Lolcommits Yammer Authorization Complete")
          else
            res.body = oauth_response("Lolcommits Authorization Cancelled")
          end
        end
      end
    end
  end
end
